python/foglamp/services/core/api/auth.py

```diff
@@ -62,35 +64,61 @@
         curl -X POST -d '{"username": "user", "password": "foglamp"}' http://localhost:8081/foglamp/login
     """
 
-    data = await request.json()
-    username = data.get('username')
-    password = data.get('password')
+    data = await request.text()
+    if data.startswith("-----BEGIN CERTIFICATE-----"):
+        peername = request.transport.get_extra_info('peername')
+        if peername is not None:
+            host, port = peername
 
-    if not username or not password:
-        _logger.warning("Username and password are required to login")
-        raise web.HTTPBadRequest(reason="Username or password is missing")
+        cert = None
+        try:
+            cert = BaseCert.from_pem(data)
+        except ValueError:
+            cert = None
+        if not cert:
+            raise web.HTTPBadRequest(reason="Invalid Certificate")
 
-    username = str(username).lower()
+        try:
+            await User.Objects.verify_certificate(cert)
+            username = cert.get_common_name()
+            uid, token, is_admin = await User.Objects.certificate_login(username, host)
+            # set the user to request object
+            request.user = await User.Objects.get(uid=uid)
+            # set the token to request
+            request.token = token
+        except (VerificationError, User.DoesNotExist) as e:
+            raise web.HTTPUnauthorized(reason="Authentication failed")
```

Comment:
It should be 404 for `User.DoesNotExist`

Reply:
In the case of certificate login, it should be HTTPUnauthorized only. Or we should find a mechanism to know username and check that first and based on that return DoesNotExist  and if exists and VerificationError then   HTTPUnauthorized.
**TO keep things simple, on login with certificate - HTTPUnauthorized is good enough, as all info is embedded in certificate only**